### PR TITLE
[#387] bump keycloak version

### LINKED ISSUE #387: Bump Keycloak to `21.0.1`
Docker changes w.r.t. https://github.com/hzi-braunschweig/SORMAS-Project/issues/8018

### PR DESCRIPTION
**Pull request by JonasCir**
fixes #387

---
### What has been changed in the `keycloak` Dockerfile?

| Subsystem   |      Before      |  > 18.0 |
|----------|:-------------:|------:|
| runtime |  WildFly | Quarkus |
| system configuration |  `configuration/standalone{-ha}.xml`    | ENV or CLI args |
| realm import | via `SORMAS.json` |    unchanged |
| templating of `SORMAS.json` |via `setup-keycloak.sh` | `$VAR` read from ENV |
|realm update | via `update-realm.sh`| `kc.sh start --import-realm` with SORMAS.json in data/import  |
|logging | `standalone-logging.cli`|`kc.sh --log-level`|
|themes|`JBOSS_HOME/themes`|`/opt/keycloak/themes`|
|SPI providers| `$JBOSS_HOME/standalone/deployments`|`/opt/keycloak/themes`|
|SPI enabled|automatic|`kc.sh build --spi-password-hashing-sormas-sha256-enabled=true`|
|DB|???|`kc.sh build  --db postgres`|
|Docker build| single stage | multi stage|

As before, Keycloak is serving via HTTP (what is discouraged as secret tokens are then present in plain text between TLS termination/reverse proxy and container).

This requires a little more setup in the new version:

`--import-realm` imports all realm JSON files provided in data/import https://www.keycloak.org/server/containers#_importing_a_realm_on_startup
`--hostname-strict-https` b/c of keycloak/keycloak#11922
`--http-enabled` / `--proxy` for HTTP between reverse proxy and the container
`--http-relative-path` to register the reverse proxy path

**Important:** The old version used to skip over resources during the import from `SORMAS.json` on conflict. The only option left in the new version is to override any existing configuration with the most recent version, as described [here](https://www.keycloak.org/server/importExport).

### CODE REVIEW

**Comment by JonasCir** [comment]
Build wont work currently as I decided to changed the import, but still good to review to get a first impression...

**Comment by JonasCir** [comment]
Unfortunately blocked on https://github.com/keycloak/keycloak/issues/12939 it seems

**Comment by JonasCir** [comment]
Also filed https://github.com/keycloak/keycloak/issues/13171, but not certain if this is important

**Comment by fhauptmann** [comment]
Keycloak 18 has been cancelled until blocking tasks have been resolved

**Comment by MartinWahnschaffe** [comment]
I have fixed the config import problem and updated keycloak to 21.0.1. All changes have also been reflected in the description above.

File: .env [hunk 1]
```diff
@@ -16,7 +14,6 @@ KEYCLOAK_DB_HOST=keycloak-postgres
 KEYCLOAK_DB_NAME=keycloak
```
Comment (MartinWahnschaffe) [inline .env:14]: Remove this, as it's no longer used

File: keycloak/update-realm.sh [hunk 2]
```diff
@@ -1,10 +0,0 @@
-#!/bin/bash
-
-KCADM="/opt/jboss/keycloak/bin/kcadm.sh"
-
-until $(curl --output /dev/null --silent --head --fail http://localhost:8080/keycloak/auth); do
-    sleep 10s
-done
-
-${KCADM} config credentials --server http://localhost:8080/keycloak/auth --user ${KEYCLOAK_USER} --password ${KEYCLOAK_PASSWORD} --realm master
-${KCADM} create partialImport -r SORMAS -s ifResourceExists=SKIP -o -f ./SORMAS.json
```
Comment (MartinWahnschaffe) [inline keycloak/update-realm.sh:10]: The functionality to add new resources to the realm by importing the SORMAS.json after startup is missing now. The `kc start --import-realm` command will skip the realm completely if it already exists.

File: services-base.yml [hunk 3]
```diff
@@ -187,7 +190,7 @@ services:
     mem_limit: ${KEYCLOAK_MEM}
     mem_reservation: ${KEYCLOAK_MEM_RESERVED}
     healthcheck:
-      test: ["CMD", "curl", "-f", "-I", "http://localhost:8080/keycloak/auth/realms/master"]
+      test: [ "CMD", "curl", "-f", "-I", "http://localhost:8080/auth/" ]
```
Comment (MartinWahnschaffe) [inline services-base.yml:193]: Does not work since curl is no longer available. Also /auth is the wrong path.

**Review by MartinWahnschaffe — CHANGES_REQUESTED** [review]

**Review by MartinWahnschaffe — APPROVED** [review]

**Comment by fhauptmann** [comment]
Successfully deployed

**Review by tnako-sg — APPROVED** [review]

**Review by fhauptmann — APPROVED** [review]

**Pull request merged**